tensorflow_quantum/python/optimizers/spsa_minimizer.py

```diff
@@ -0,0 +1,275 @@
+        'objective_value',
+        # A tensor containing the value of the objective
+        # function at the `position`. If the search
+        # converged, then this is the (local) minimum of
+        # the objective function.
+        'tolerance',
+        # Define the stop criteria. Iteration will stop when the
+        # objective value difference between two iterations is
+        # smaller than tolerance
+        'a',
+        # Specifies the learning rate
+        'alpha',
+        # Specifies scaling of the learning rate
+        'c',
+        # Specifies the size of the perturbations
+        'gamma',
+        # Specifies scaling of the size of the perturbations
+        'blocking',
+        # If true, then the optimizer will only accept updates that improve
+        # the objective function.
+        'allowed_increase'
+        # Specifies maximum allowable increase in objective function
+        # (only applies if blocking is true).
+    ])
+
+
+def _get_initial_state(initial_position, tolerance, expectation_value_function,
+                       a, alpha, c, gamma, blocking, allowed_increase):
+    """Create SPSAOptimizerResults with initial state of search."""
+    init_args = {
+        "converged": tf.Variable(False),
+        "num_iterations": tf.Variable(0),
+        "num_objective_evaluations": tf.Variable(0),
+        "position": tf.Variable(initial_position),
+        "objective_value": tf.Variable(0.),
+        "objective_value_previous_iteration": tf.Variable(0.),
+        "tolerance": tolerance,
+        "a": tf.Variable(a),
+        "alpha": tf.Variable(alpha),
+        "c": tf.Variable(c),
+        "gamma": tf.Variable(gamma),
+        "blocking": tf.Variable(blocking),
+        "allowed_increase": tf.Variable(allowed_increase)
+    }
+    return SPSAOptimizerResults(**init_args)
+
+
+def minimize(expectation_value_function,
+             initial_position,
+             tolerance=1e-7,
+             max_iterations=60,
+             alpha=0.602,
+             a=1.0,
+             c=1.0,
+             gamma=0.101,
+             blocking=True,
+             allowed_increase=0.5,
+             name=None):
+    """Applies the SPSA algorithm.
+
+    The SPSA algorithm can be used to minimize a noisy function. See:
+
+    [SPSA website](https://www.jhuapl.edu/SPSA/)
+
+    Usage:
+
+    Here is an example of optimize a function which consists summation of
+    a few sinusoids.
+
+    >>> n = 10  # Number of sinusoids
+    >>> coefficient = tf.random.uniform(shape=[n])
+    >>> min_value = -tf.math.reduce_sum(tf.abs(coefficient))
+    >>> func = lambda x:tf.math.reduce_sum(tf.sin(x) * coefficient)
+    >>> # Optimize the function with SPSA, start with random parameters
+    >>> result = tfq.optimizers.SPSA_minimize(func, np.random.random(n))
+    >>> result.converged
+    tf.Tensor(True, shape=(), dtype=bool)
+    >>> result.objective_value
+    tf.Tensor(-4.7045116, shape=(), dtype=float32)
```

Comment:
Can we add a unit test that covers the use case shown here ?

Reply:
I modified the example and added the same format test case